android/compose/src/main/kotlin/com/boswelja/ephemeris/compose/EphemerisCalendarState.kt

```diff
@@ -0,0 +1,105 @@
+import androidx.compose.runtime.Stable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.setValue
+import androidx.compose.runtime.snapshotFlow
+import com.boswelja.ephemeris.core.data.CalendarPageSource
+import com.boswelja.ephemeris.core.ui.CalendarPageLoader
+import com.boswelja.ephemeris.core.ui.CalendarState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+import kotlinx.datetime.LocalDate
+
+/**
+ * A base implementation of [CalendarState] designed to be used with [EphemerisCalendar].
+ */
+public abstract class EphemerisCalendarState : CalendarState {
+    /**
+     * The internal [InfinitePagerState] that [EphemerisCalendar] uses to control its page state.
+     */
+    internal abstract val pagerState: InfinitePagerState
+
+    /**
+     * The internal [CalendarPageLoader] that [EphemerisCalendar] uses to load page data. This should
+     * be backed by a Composable State so the calendar can react to changes properly.
+     */
+    internal abstract var pageLoader: CalendarPageLoader
+}
+
+/**
+ * The default implementation of [EphemerisCalendarState]. This is returned by [rememberCalendarState]
+ * @param pagerState The default [CalendarPageSource] to use. This can be changed later.
+ * @param coroutineScope A coroutine scope to use for running Calendar-related jobs.
+ */
+internal class EphemerisCalendarStateImpl internal constructor(
+    calendarPageSource: CalendarPageSource,
+    private val coroutineScope: CoroutineScope,
+    override val pagerState: InfinitePagerState
+) : EphemerisCalendarState() {
+
+    override var pageLoader by mutableStateOf(CalendarPageLoader(coroutineScope, calendarPageSource))
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override val displayedDateRange: StateFlow<ClosedRange<LocalDate>> = snapshotFlow { pageLoader }
+        .flatMapLatest { snapshotFlow { pagerState.page } }
+        .map { pageLoader.getDateRangeFor(it) }
+        .stateIn(
+            coroutineScope,
+            SharingStarted.Eagerly,
+            pageLoader.getDateRangeFor(pagerState.page)
+        )
+
+    override var pageSource: CalendarPageSource = calendarPageSource
+        set(value) {
+            if (field != value) {
+                field = value
+                pageLoader = CalendarPageLoader(
+                    coroutineScope,
+                    value
+                )
+            }
+        }
+
+    override fun scrollToDate(date: LocalDate) {
+        // We launch this in a coroutine to avoid any thread blocking
+        coroutineScope.launch {
```

Comment:
might need to check that the scope is active before trying this, especially as it can be any scope, not specifically a `rememberCoroutineScope`

Reply:
What would the expected behavior be if the scope isn't active? Or would you rather not allow users to pass a scope?

Reply:
it'd be a no-op if it's cancelled, I don't know about `rememberCoroutineScope`, but if it's tied to the lifecycle of the Composable, it's a safer way to ensure that it'll be launched without any of the aforementioned side effects and the user won't see anything happen

Reply:
Yep, `rememberCoroutineScope` is tied to the lifecycle of the call site. I think not allowing users to pass a coroutine scope will work fine in this case